docs/madr/decisions/017-mesh-proxy-template.md

```diff
@@ -0,0 +1,244 @@
+# MeshProxyTemplate
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4737
+
+## Context and Problem Statement
+
+We need to revamp ProxyTemplate for the new policy matching.
+Proxy Template lets you adjust Envoy Config when Kuma policies do not provide a native way to configure thing you want.
+In a secured setup, it's only safe for mesh operators to operate Proxy Template. Otherwise, service owners could bypass any policy.
+
+There are a couple of problems with existing ProxyTemplate policy 
+
+### No aggregation
+
+Let say that there is a policy that selects all DP proxies and tweaks one value.
+Then, if there is another ProxyTemplate that matches specific DPP, the global one is not applied anymore.
+
+### Profile imports section
+
+The policy contains `imports` section which is not really useful.
+Currently, you need to specify a profile in `imports`, otherwise the default config won't be generated.
+Available values are `default-profile` or `gateway-profile`.
+
+One use case is to skip default config generation and build your own config by scratch.
+However, we haven't seen this in the wild. You can always apply modifications that delete all the clusters, listeners etc.
+
+Additionally, the requirement of profile for a default config is a blocker when you want to apply mesh wide proxy template (`kuma.io/service: '*'`), but you have both sidecar and gateway proxies.
+
+When we created we were predicting that we may have user-created profiles.
+This never happened in the end. We only used profiles internally.
+If someone would like to build their own config I presume they would rather implement this as a plugin. Maintaing whole Envoy config in ProxyTemplate is not really maintainable solution.
+If we ever see a need to skip the default config, we can do this by introducing a new flag like `skipDefaultConfigGeneration` in `defaults` section.
+
+### No modifications on zone proxies
+
+Since ProxyTemplate is a mesh-wide policy, we cannot modify zone proxies.
+Solving this would require duplicating ProxyTemplate policy with global scope. This won't be covered by this MADR.
```

Comment:
do we have an issue for this?

Reply:
Yeah let's open an issue :)

Reply:
https://github.com/kumahq/kuma/issues/5558